### PR TITLE
Upgrade to rails 5

### PR DESCRIPTION
**Pull request by h-lame**
Best reviewed commit-by-commit.

Still one dep warning at startup for kaminari, but doing the jump to 1.0 to avoid doesn't seem worth it.

For: https://trello.com/c/u8hJdr3O/161-migrate-to-rails-5-need-api

### CODE REVIEW

File: app/models/need.rb [hunk 1]
```diff
@@ -166,11 +166,19 @@ def validate_duplicate
...
+    end
   end
 
   def record_update_revision
-    record_revision "update"
+    if @during_record_revision_callback.nil?
+      @during_record_revision_callback = 'yes'
+      record_revision "update"
+      @during_record_revision_callback = nil
+    end
```
Comment (tuzz) [inline app/models/need.rb]: Perhaps we could wrap this behaviour?

```
prevent_additional_callbacks { record_revision "create" }
```

File: app/models/need_revision.rb [hunk 2]
```diff
@@ -8,7 +8,7 @@ class NeedRevision
 
   default_scope -> { order_by(:created_at.desc) }
 
-  belongs_to :need
+  belongs_to :need, touch: false, inverse_of: :revisions
```
Comment (tuzz) [inline app/models/need_revision.rb]: If you take the `touch: false` away, does that re-introduce the problem?

File: Gemfile [hunk 3]
```diff
@@ -34,7 +34,7 @@ else
   gem 'gds-sso', '13.2.0'
 end
 
-gem 'airbrake', '~> 4.3.4'
+gem 'airbrake', github: 'alphagov/airbrake', branch: 'silence-dep-warnings-for-rails-5'
```
Comment (tuzz) [inline Gemfile]: Do you know if the intention is for this branch to stay around for a while? I wonder if we could wrap its silencing behaviour in something that checks if the Rails version is 5 and then merge it into master?

This is probably out of scope for this upgrade, though.

**Review by tuzz — APPROVED** [review]
A few minor suggestions. LGTM 👍

**Comment by tuzz** [comment]
FYI, I think you rebase-pushed so a couple of comments are showing as 'outdated'

File: app/models/need.rb (hunk 1 above)
Comment (h-lame) [inline app/models/need.rb]: Oh, yes; good call.  Makes it more obvious, will do.

File: app/models/need_revision.rb (hunk 2 above)
Comment (h-lame) [inline app/models/need_revision.rb]: TBH, I dunno.  I'll try it and see (I was trying all the things in one go).  It'll be better to introduce the smallest change that avoids the problem.

File: Gemfile (hunk 3 above)
Comment (h-lame) [inline Gemfile]: I think it's a stopgap until we move off errbit and onto something else entirely.  Be totally worth having this discussion in a wider forum though.

**Comment by h-lame** [comment]
@tuzz fixed what you said in a rebase; if you want to re-look.

File: app/models/need_revision.rb (hunk 2 above)
Comment (h-lame) [inline app/models/need_revision.rb]: Done in a rebase - it all still worked without `touch: false` so I dropped it.

File: app/models/need.rb (hunk 1 above)
Comment (h-lame) [inline app/models/need.rb]: Done in a rebase.

**Comment by h-lame** [comment]
Tested the branch on integration and checked that whitehall can still show needs in the admin.  I can't test metadata-api as it's broken.

**Pull request merged**